dev/src/cubegen.cpp

```diff
@@ -559,134 +559,160 @@ nfr("cg_load_vox", "name", "S", "R:voxels]S?",
+        if ((bufs.size() >= 8) && (strncmp((const char *)bufs.data(), "VOX ", 4) == 0)) {
             // This looks like a MagicaVoxel file.
             int3 size = int3_0;
-            auto p = buf.c_str() + 8;
+            bufs = bufs.subspan(8);
             bool chunks_skipped = false;
             Voxels *voxels = nullptr;
             map<int32_t, int32_t> node_graph;
             map<int32_t, int32_t> node_to_model;
             map<int32_t, int32_t> node_to_layer;
             map<int32_t, string> layer_names;
             map<int32_t, string> node_names;
-            while (p < buf.c_str() + buf.length()) {
-                auto id = p;
-                p += 4;
-                auto contentlen = *((int *)p);
-                p += 8;
+
+            while (bufs.size() >= 8) {
+                auto id = (const char *)bufs.data();
+                bufs =bufs.subspan(4);
+                int contentlen;
+                if (!ReadSpanInc(bufs, contentlen)) return erreof();
+                bufs= bufs.subspan(4);
+                if (bufs.size() < contentlen) return erreof();
+                auto p = bufs.subspan(0, contentlen);
+                bufs = bufs.subspan(contentlen);
                 if (!strncmp(id, "SIZE", 4)) {
-                    size = int3((int *)p);
+                    if (!ReadSpanInc(p, size)) return erreof();
                     voxels = NewWorld(size, default_palette_idx);
                     voxvec->Push(vm, Value(vm.NewResource(&voxel_type, voxels)));
                 } else if (!strncmp(id, "RGBA", 4)) {
                     if (!voxels) return errf(".vox file RGBA chunk in wrong order");
                     vector<byte4> palette;
                     palette.push_back(byte4_0);
-                    palette.insert(palette.end(), (byte4 *)p, ((byte4 *)p) + 255);
+                    if (p.size() < 256) return erreof();
+                    palette.insert(palette.end(), (byte4 *)p.data(), ((byte4 *)p.data()) + 255);
                     auto pi = NewPalette(palette.data());
                     for (iint i = 0; i < voxvec->len; i++) {
                         GetVoxels(voxvec->At(i)).palette_idx = pi;
                     }
                 } else if (!strncmp(id, "XYZI", 4)) {
                     if (!voxels) return errf(".vox file XYZI chunk in wrong order");
-                    auto numvoxels = *((int *)p);
+                    int numvoxels;
+                    if (!ReadSpanInc(p, numvoxels)) return erreof();
+                    if (p.size_bytes() < numvoxels * sizeof(byte4)) return erreof();
+                    auto vp = (uint8_t *)p.data();
                     for (int i = 0; i < numvoxels; i++) {
-                        auto vox = byte4((uint8_t *)(p + i * 4 + 4));
+                        auto vox = byte4((vp + i * 4));
                         auto pos = int3(vox.xyz());
                         if (pos < voxels->grid.dim) voxels->grid.Get(pos) = vox.w;
                     }
+
```

Comment:
I think I just removed these?

Reply:
what do you mean ?

Reply:
you're adding blank lines that I just removed..